### ISSUE TITLE
Topic Filter - RangeError (index) with single-level wildcard

### ISSUE THREAD

**Issue by Knechtie** [shamblett/mqtt_client#85]
When creating a topic filter with a single-level wildcard (+) at the beginning, the following error occurs:

```
E/flutter ( 4194): [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: RangeError (index): Invalid value: Only valid value is 0: 1
E/flutter ( 4194): #0      List.[] (dart:core/runtime/libgrowable_array.dart:145:60)
E/flutter ( 4194): #1      SubscriptionTopic.matches 
../…/src/mqtt_client_subscription_topic.dart:84
E/flutter ( 4194): #2      MqttClientTopicFilter._topicIn 
../…/management/mqtt_client_topic_filter.dart:44
E/flutter ( 4194): #3      _rootRunUnary (dart:async/zone.dart:1132:38)
E/flutter ( 4194): #4      _CustomZone.runUnary (dart:async/zone.dart:1029:19)
E/flutter ( 4194): #5      _CustomZone.runUnaryGuarded (dart:async/zone.dart:931:7)
E/flutter ( 4194): #6      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:336:11)
E/flutter ( 4194): #7      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:263:7)
E/flutter ( 4194): #8      _SyncBroadcastStreamController._sendData.<anonymous closure> (dart:async/broadcast_stream_controller.dart:383:20)
E/flutter ( 4194): #9      _BroadcastStreamController._forEachListener (dart:async/broadcast_stream_controller.dart:324:15)
E/flutter ( 4194): #10     _SyncBroadcastStreamController._sendData (dart:async/broadcast_stream_controller.dart:382:5)
E/flutter ( 4194): #11     _BroadcastStreamController.add (dart:async/broadcast_stream_controller.dart:250:5)
E/flutter ( 4194): #12     ChangeNotifier.deliverChanges 
package:mqtt_client/…/src/change_notifier.dart:61
E/flutter ( 4194): #13     _rootRun (dart:async/zone.dart:1120:38)
E/flutter ( 4194): #14     _CustomZone.run (dart:async/zone.dart:1021:19)
E/flutter ( 4194): #15     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
E/flutter ( 4194): #16     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
E/flutter ( 4194): #17     _rootRun (dart:async/zone.dart:1124:13)
E/flutter ( 4194): #18     _CustomZone.run (dart:async/zone.dart:1021:19)
E/flutter ( 4194): #19     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
E/flutter ( 4194): #20     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
E/flutter ( 4194): #21     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
E/flutter ( 4194): #22     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
```

The corresponding code:
```Dart
    final MqttClientTopicFilter topicFilterSystemStructure =  MqttClientTopicFilter('+/systemStructure', client.updates);
```

**Comment by shamblett** [comment]
OK, do you know what your incoming topic string was? If I try your topic filer with a topic like 1/systemStructure or 101/systemStructure it works OK.

**Comment by Knechtie** [comment]
Sorry for the vague information in my first post. I'm pretty new to Dart and MQTT.

I found out that this error not only occurs by using the single-level wildcard. Instead it seems that it appears when I use multiple topic filters.

Here are some cases:

```
Topic: "PLCs/numberOfSystems"
Payload: "1"
QoS: "1"
```
**No Error:**
``` Dart
    const String topicNumberOfSystems = 'PLCs/numberOfSystems';
    const String topicSystemStructure = 'PLCs/+/systemStructure';

    //await Future.delayed(Duration(seconds: 3), () {});
    // Topicfilter erzeugen
    client.subscribe(topicNumberOfSystems, MqttQos.exactlyOnce);
    // client.subscribe(topicSystemStructure, MqttQos.exactlyOnce);

    //  final MqttClientTopicFilter topicFilterSystemStructure =
    //     MqttClientTopicFilter(topicSystemStructure, client.updates);

    final MqttClientTopicFilter topicFilterNumberOfSystems =
        MqttClientTopicFilter(topicNumberOfSystems, client.updates);
```
**Error:**
``` Dart
    const String topicNumberOfSystems = 'PLCs/numberOfSystems';
    const String topicSystemStructure = 'PLCs/+/systemStructure';

    //await Future.delayed(Duration(seconds: 3), () {});
    // Topicfilter erzeugen
    client.subscribe(topicNumberOfSystems, MqttQos.exactlyOnce);
    // client.subscribe(topicSystemStructure, MqttQos.exactlyOnce);

    final MqttClientTopicFilter topicFilterSystemStructure =
        MqttClientTopicFilter(topicSystemStructure, client.updates);

    final MqttClientTopicFilter topicFilterNumberOfSystems =
        MqttClientTopicFilter(topicNumberOfSystems, client.updates);

```

``` Dart

```

``` Dart

```

**Comment by Knechtie** [comment]
Sorry for the vague information in my first post. I'm new to Dart and MQTT.

I found out that this error not only occurs when using the single-level wildcard. Instead it seems that it occurs when I use multiple topic filters and receive a message with two topic levels (e.g. PLCs/numberOfSystems).

Here are some cases:

### Topic: PLCs/numberOfSystems
```
Published Message:
   Topic: "PLCs/numberOfSystems"
   Payload: "1"
   QoS: "1"
   Retain: "true"
```
**No error (one filter):**
``` Dart
    const String topicNumberOfSystems = 'PLCs/numberOfSystems';
    const String topicSystemStructure = 'PLCs/+/systemStructure';

    client.subscribe(topicNumberOfSystems, MqttQos.exactlyOnce);
    // client.subscribe(topicSystemStructure, MqttQos.exactlyOnce);

    //  final MqttClientTopicFilter topicFilterSystemStructure =
    //     MqttClientTopicFilter(topicSystemStructure, client.updates);

    final MqttClientTopicFilter topicFilterNumberOfSystems =
        MqttClientTopicFilter(topicNumberOfSystems, client.updates);
```
**Error (two filters):**
``` Dart
    const String topicNumberOfSystems = 'PLCs/numberOfSystems';
    const String topicSystemStructure = 'PLCs/+/systemStructure';

    client.subscribe(topicNumberOfSystems, MqttQos.exactlyOnce);
    // client.subscribe(topicSystemStructure, MqttQos.exactlyOnce);

    final MqttClientTopicFilter topicFilterSystemStructure =
        MqttClientTopicFilter(topicSystemStructure, client.updates);

    final MqttClientTopicFilter topicFilterNumberOfSystems =
        MqttClientTopicFilter(topicNumberOfSystems, client.updates);

```
---
### Topic: PLCs/test/numberOfSystems
```
Published Message:
   Topic: "PLCs/test/numberOfSystems"
   Payload: "1"
   QoS: "1"
   Retain: "true"
```
**No error (two filters):**
``` Dart
    const String topicNumberOfSystems = 'PLCs/test/numberOfSystems';
    const String topicSystemStructure = 'PLCs/+/systemStructure';

    client.subscribe(topicNumberOfSystems, MqttQos.exactlyOnce);
    // client.subscribe(topicSystemStructure, MqttQos.exactlyOnce);

    final MqttClientTopicFilter topicFilterSystemStructure =
        MqttClientTopicFilter(topicSystemStructure, client.updates);

    final MqttClientTopicFilter topicFilterNumberOfSystems =
        MqttClientTopicFilter(topicNumberOfSystems, client.updates);
```
---
### Topic: PLCs/test
```
Published Message:
   Topic: "PLCs/test"
   Payload: "1"
   QoS: "1"
   Retain: "true"
```
**Error (two filters):**
``` Dart
    const String topicNumberOfSystems = 'PLCs/test';
    const String topicSystemStructure = 'PLCs/+/systemStructure';

    client.subscribe(topicNumberOfSystems, MqttQos.exactlyOnce);
    // client.subscribe(topicSystemStructure, MqttQos.exactlyOnce);

    final MqttClientTopicFilter topicFilterSystemStructure =
        MqttClientTopicFilter(topicSystemStructure, client.updates);

    final MqttClientTopicFilter topicFilterNumberOfSystems =
        MqttClientTopicFilter(topicNumberOfSystems, client.updates);
```

**Log of this case**

```
I/flutter ( 3790): MQTT client connected
I/flutter ( 3790): 2019-03-27 16:03:21.450312 -- MqttConnectionHandler::sendMessage - MQTTMessage of type MqttMessageType.subscribe
I/flutter ( 3790): Header: MessageType = MqttMessageType.subscribe, Duplicate = false, Retain = false, Qos = MqttQos.atLeastOnce, Size = 0
I/flutter ( 3790): Subscribe Variable Header: MessageIdentifier={1}
I/flutter ( 3790): Payload: Subscription [{1}]
I/flutter ( 3790): {{ Topic={PLCs/test}, Qos={MqttQos.exactlyOnce} }}
I/flutter ( 3790): 
I/flutter ( 3790): 2019-03-27 16:03:21.468380 -- MqttConnection::_onData
I/flutter ( 3790): 2019-03-27 16:03:21.479021 -- MqttConnection::_onData - message received MQTTMessage of type MqttMessageType.subscribeAck
I/flutter ( 3790): Header: MessageType = MqttMessageType.subscribeAck, Duplicate = false, Retain = false, Qos = MqttQos.atMostOnce, Size = 3
I/flutter ( 3790): SubscribeAck Variable Header: MessageIdentifier={1}
I/flutter ( 3790): Payload: Qos grants [{1}]
I/flutter ( 3790): {{ Grant={MqttQos.exactlyOnce} }}
I/flutter ( 3790): 
I/flutter ( 3790): 2019-03-27 16:03:21.479408 -- MqttConnection::_onData - message processed
I/flutter ( 3790): 2019-03-27 16:03:21.496559 -- MqttConnection::_onData - message received MQTTMessage of type MqttMessageType.publish
I/flutter ( 3790): Header: MessageType = MqttMessageType.publish, Duplicate = false, Retain = true, Qos = MqttQos.atLeastOnce, Size = 14
I/flutter ( 3790): Publish Variable Header: TopicName={PLCs/test}, MessageIdentifier={1}, VH Length={13}
I/flutter ( 3790): Payload: {1 bytes={<49>
I/flutter ( 3790): 2019-03-27 16:03:21.497446 -- MqttConnection::_onData - message processed
I/flutter ( 3790): 2019-03-27 16:03:21.518539 -- MqttConnectionHandler::sendMessage - MQTTMessage of type MqttMessageType.publishAck
I/flutter ( 3790): Header: MessageType = MqttMessageType.publishAck, Duplicate = false, Retain = false, Qos = MqttQos.atMostOnce, Size = 0
I/flutter ( 3790): PublishAck Variable Header: MessageIdentifier={1}
I/flutter ( 3790): 1

E/flutter ( 3790): [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: RangeError (index): Invalid value: Not in range 0..1, inclusive: 2
E/flutter ( 3790): #0      List.[] (dart:core/runtime/libgrowable_array.dart:145:60)
E/flutter ( 3790): #1      SubscriptionTopic.matches 
../…/src/mqtt_client_subscription_topic.dart:84
E/flutter ( 3790): #2      MqttClientTopicFilter._topicIn 
../…/management/mqtt_client_topic_filter.dart:44
E/flutter ( 3790): #3      _rootRunUnary (dart:async/zone.dart:1132:38)
E/flutter ( 3790): #4      _CustomZone.runUnary (dart:async/zone.dart:1029:19)
E/flutter ( 3790): #5      _CustomZone.runUnaryGuarded (dart:async/zone.dart:931:7)
E/flutter ( 3790): #6      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:336:11)
E/flutter ( 3790): #7      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:263:7)
E/flutter ( 3790): #8      _SyncBroadcastStreamController._sendData.<anonymous closure> (dart:async/broadcast_stream_controller.dart:383:20)
E/flutter ( 3790): #9      _BroadcastStreamController._forEachListener (dart:async/broadcast_stream_controller.dart:324:15)
E/flutter ( 3790): #10     _SyncBroadcastStreamController._sendData (dart:async/broadcast_stream_controller.dart:382:5)
E/flutter ( 3790): #11     _BroadcastStreamController.add (dart:async/broadcast_stream_controller.dart:250:5)
E/flutter ( 3790): #12     ChangeNotifier.deliverChanges 
package:mqtt_client/…/src/change_notifier.dart:61
E/flutter ( 3790): #13     _rootRun (dart:async/zone.dart:1120:38)
E/flutter ( 3790): #14     _CustomZone.run (dart:async/zone.dart:1021:19)
E/flutter ( 3790): #15     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
E/flutter ( 3790): #16     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
E/flutter ( 3790): #17     _rootRun (dart:async/zone.dart:1124:13)
E/flutter ( 3790): #18     _CustomZone.run (dart:async/zone.dart:1021:19)
E/flutter ( 3790): #19     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
E/flutter ( 3790): #20     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
E/flutter ( 3790): #21     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
E/flutter ( 3790): #22     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
E/flutter ( 3790): 
```

**Comment by shamblett** [comment]
Thanks for the extra info, I'll have a look at this.

**Comment by shamblett** [comment]
OK, I should have got to this earlier, from your example above 

final MqttClientTopicFilter topicFilterSystemStructure =
        MqttClientTopicFilter(topicSystemStructure, client.updates);

where topicSystemStructure is 'PLCs/+/systemStructure';

won't work, you can only filter on receivable topics, you can't receive a wildcard topic, only subscribe to it. That said error recovery should be more graceful than whats happening, I've updated the code to fail the match if we generate an exception and updated the API comment for the topic filter class.

**Comment by shamblett** [comment]
Client updated as above, re-published at 5.5.3
